WIP: Meson build

whs opened the pull request:
Per my offer on Mastodon, here's a  Meson build implementation for libdatrie. Hopefully closes #13, as it should be less painful than autotools but it's not CMake.

What works:

- Building on Arch Linux
- Building tools
- Installing, including manpages and doxygen

Needs to know for maintainer:

- Version needs to be bumped in `/meson.build` line 1. This version is used as .so suffix.
- There is no support for `-liconv` detection. Instead, it would try building without flags, and then use [meson dependency resolution](https://mesonbuild.com/Dependencies.html#dependency-detection-method) to find iconv.

What needs to be tested:

- Building on Mac (I have a loaned Mac, so I'll test and update)
- Building on other non-Linux (hopefully we would just ship dual meson-autoconf build and wait for feedback) 
- I'm not sure if linker versioning & exports are correct, as meson doesn't have the same feature as autoconf and I don't know that much about linker

---

What doesn't work yet:

- Cutting release tar (I'll look into this)
- Generate pkgconfig (I'll also look into this)

whs commented:
```
[8/43] Compiling C object tests/test_walk.p/test_walk.c.o
../tests/test_walk.c:78:26: warning: format specifies type 'wint_t' (aka 'int') but the argument has type 'AlphaChar' (aka 'unsigned int') [-Wformat]
        printf ("'%lc'", walkables[i]);
                  ~~~    ^~~~~~~~~~~~
                  %c

```

Got this on Mac

```
CPU: quad-core 64-bit kabylake
Clang: 12.0.5 build 1205
Git: 2.32.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.5.2-x86_64
CLT: 12.5.1.0.1.1623191612
Xcode: 12.5.1
```

All tests were passing though

pepa65 commented:
Doesn't sound like an improvement over the old system (so far at least)...  I know 1 person complained, but for everyone else autotools seems to work OK.

eli-schwartz reviewed (commented):
Hi, nice to see you considering meson. Some suggestions for potential optimizations:

whs commented on meson.build:
```diff
@@ -0,0 +1,80 @@
+deps = []
+
+if fs.exists('VERSION')
+    version = run_command(cat, files('VERSION'), check : true).stdout().strip()
```
Debian stable has meson at 0.56.2, so I don't think I should go more than that. (I tried targeting 0.49 for buster, but the lack of fs module make this infeasible)

eli-schwartz replied:
Ah, I was afraid of that. :(

whs commented on meson.build:
```diff
@@ -0,0 +1,80 @@
+
+# Detect for iconv
+compiler_has_iconv = compiler.has_function('iconv_open')
+iconv_pkgconfig = get_option('iconv_pkgconfig')
```
I think the flag is misnamed, it's more like use Meson's resolver to find iconv (which is required on a Mac - FindIconv is needed)

I think it should be harmless to remove the flag entirely and always use the `auto` path here?

eli-schwartz replied:
Does FindIconv do anything above and beyond detecting the existence of a `find_library('iconv')`?

Because having this detection not work on systems without cmake installed, is unfortunate.

whs replied:
Since iconv doesn't come with pkg-config, and we only check for `iconv_open` here, it doesn't work on a Mac which needs `-liconv` to link.  FindIconv does give us the flags we need.

eli-schwartz replied:
Right, but `compiler.find_library('iconv')` also gives you `-liconv` without using cmake to look it up.

whs replied:
Turns out [mingw-w64 does ship iconv](https://packages.msys2.org/package/mingw-w64-x86_64-libiconv). I guess I have to add back dependency support.

eli-schwartz replied:
That pkg-config file appears to come from the msys2 packaging, not upstream iconv. And it returns the same information that find_library() returns.

I would not bother using it since it's not an official pkg-config file.

whs commented:
Updated

- Removed `iconv_pkgconfig` and add check with `compiler.find_library`
- Removed `ld_no_undefined`

whs commented on datrie/meson.build:
```diff
@@ -0,0 +1,77 @@
+    link_depends : link_depends,
+    version : meson.project_version(),
+    include_directories : configuration_inc,
+    vs_module_defs : 'libdatrie.def',
```
This break Windows gcc

eli-schwartz replied:
The vs module defs file? Why?

whs replied:
I thought it would supply the .def file generated in the block for msvc above (I think I forgot to change this line to reference the generated file). However, the original def file here got passed to gcc which breaks.

I'm not sure why - I'm not exactly familiar with msvc/mingw (and it seems this project doesn't run on Windows either) I'm guessing that both compilers use the same linker def format on Windows? Maybe if I supply the generated file to gcc it might work.

eli-schwartz replied:
Ah, right, you'll need the `link_def` variable reference which contains a custom_target output.

I'm not actually sure I understand the custom_target? Is it just prepending a single constant line? Why isn't this changed in the source file so that you don't need to generate anything?

Also yes, the GNU compiler implementation in meson passes the defs file to the linker as this is valid there.

whs replied:
Yes it is passing a single line. I'm not sure how the original file is used, but it was used as-is in autotools (and I don't know which platform it was for - on Linux it use .map) so I didn't want to touch it. On Windows you need a section marker so I took inspiration from #15 to add it during build time.

Kreijstal commented:
Thank you this solved my problem, with #29

thep commented:
First of all, thank you very much for your work, and sorry for a very slow response. It took me a long while to find some time to study meson so I can understand your proposal.

Here are what I noticed:

- Warnings from meson:
```
WARNING: Project specifies a minimum meson_version '>=0.53.0' but uses features which were added in newer versions:
 * 0.55.0: {'Passing executable/found program object to script parameter of add_dist_script', 'Passing executable/found program object to script parameter of add_install_script'}
NOTICE: Future-deprecated features used:
 * 0.56.0: {'meson.build_root', 'meson.source_root'}
```

- Library ABI version is taken from the hard-coded project version `1.4.0`, which is inappropriate, as "`meson dist`" produces `1.4.0` tarball, while the actual version is still `0.2.13`. The correct numbers should be this way:
  - Project version should be taken from the `VERSION` file generated by the `build-aux/git-version-gen` script,
  - `library()` version should be calculated from `LT_CURRENT`, `LT_REVISION` and `LT_AGE` defined in `configure.ac` like this:
```python
soversion = LT_CURRENT - LT_AGE
libversion = '@0@.@1@.@2@'.format(soversion, LT_AGE, LT_REVISION)
```

- Tarball generated by "`meson dist`" includes excessive files like `.cvsignore`, `.gitignore`, `autogen.sh`, the whole `autom4te.cache/` directory, `nsis/` directory.

When the meson build is ready, I think we can move away from autotools and maintain only one build system in the end.

eli-schwartz commented:
> * Tarball generated by "`meson dist`" includes excessive files like `.cvsignore`, `.gitignore`, `autogen.sh`, the whole `autom4te.cache/` directory, `nsis/` directory.

It includes the output of running autogen.sh which is arguably important to ensure that `meson dist` is a suitable replacement for autotools `make dist` during a transition phase from one to the other.

Arguably .cvsignore could just be deleted from the repo ;) and the gitignore file can be excluded using gitattributes export-ignore.

thep commented:
> > * Tarball generated by "`meson dist`" includes excessive files like `.cvsignore`, `.gitignore`, `autogen.sh`, the whole `autom4te.cache/` directory, `nsis/` directory.
> 
> It includes the output of running autogen.sh which is arguably important to ensure that `meson dist` is a suitable replacement for autotools `make dist` during a transition phase from one to the other.

`autogen.sh` has never been distributed with the tarball so far. It's intended for use in the repo only. Surely the cache dir `autom4te.cache/` is not for distribution. For `nsis/`, it's experimental Windows installer, which is not ready for distribution yet.

The decent outputs of `autogen.sh` not mentioned above, such as the `configure` script, `Makefile.in`, `config.guess`, `config.sub`, the `m4/` directory, etc, are not part of the problem here.

> Arguably .cvsignore could just be deleted from the repo ;) and the gitignore file can be excluded using gitattributes export-ignore.

No doubt for this.

whs commented:
So what's left for this PR to be merged? I might be able to work on it right away but would be great if I could have a list. From what I understand:

1. Bump Meson version. The current version was chosen because it was the oldest version available in major Linux distros at thetime, but it's too old to use even back then.
   - Survey for a new Meson version to use
   - Fix any deprecated features
2. Figure out how to separate library version and so version.
   - I think Meson is against running script as part of the process formerly done in `configure` phase, so running `git-version-gen` might not be possible unless Meson actually provide that exact functionality in the stdlib
3. Remove `.cvsignore`, add `.gitignore`, `autogen.sh`, `autom4te.cache`, `nsis` to .gitattribute

I don't know about the current state of Windows builds so I'm not sure if the discussion with @eli-schwartz above is still valid or not. Personally I don't want to hand test all the OSes again (I don't have access to any Mac now) so I'd prefer those test to be done by GitHub Actions. I could contribute one later if this get merged.

eli-schwartz commented:
> I think Meson is against running script as part of the process formerly done in configure phase, so running `git-version-gen` might not be possible unless Meson actually provide that exact functionality in the stdlib

There's nothing wrong with doing this.

```meson
project(...,
    version: run_command('./git-version-gen', check: true).stdout().strip()
```

Note that it will only be run on configure, so if the output changes after a git pull a new configure won't be done and the version number becomes outdated.

vcs_tag() can be used to generate a version.h file as a build edge (produced every time you run ninja) if that's an unacceptable shortcoming.

whs commented:
Updated

1. Project version now use git-version-gen and so version is now computed in meson. This means it has to be updated in both configure.ac and meson.build, but once autotools is removed it will only be in meson.
2. Dist file should be equal now. The m4 build cache was in there because I run `autogen.sh` during `dist` and unlike `autotools dist` it wasn't cleaned automatically.
3. There was a bug during dist that it do not honor DESTDIR. This is now fixed
4. Now that Meson has new iconv detection, remove special iconv logic. The other two detection I'm not exactly sure if required so I still leave it in.
5. Minimum Meson version is now 1.0. This should be minimum version for current version of all major LTS distros except RHEL9 derivatives.

I also tested the build with [muon](https://muon.build/) and the test passes. It doesn't seems to implement the same iconv check as meson, but at least on Linux-glibc one doesn't need special flags for iconv anyway.

pepa65 commented:
I am on the latest Linux Mint, it has Meson 0.61.2... But soon I hope there will be a new release.

eli-schwartz commented on man/meson.build:
```diff
@@ -1,4 +1,3 @@
 
-mandir_abs = get_option('prefix') / get_option('mandir')
-meson.add_install_script(sh, '-c', 'cd @0@ && ln -s trietool.1 trietool-0.2.1 || true'.format(mandir_abs / 'man1'))
+meson.add_install_script(sh, '-c', 'cd ${MESON_INSTALL_DESTDIR_PREFIX}/@0@/man1 && ln -s trietool.1 trietool-0.2.1 || true'.format(get_option('mandir')))
```
Since the minimum version was bumped from 0.53, and is now >=0.61.0, you can use:

```suggestion
install_symlink('trietool-0.2.1', pointing_to: 'trietool.1', install_dir: get_option('mandir') / 'man1')
```

eli-schwartz commented on meson.build:
```diff
@@ -47,7 +48,7 @@ endif
 # Detect for version-script support
 ld_has_version_script = compiler.has_link_argument(
-    '-Wl,-version-script,@0@'.format(meson.source_root() / 'build-aux' / 'test.map')
+    '-Wl,-version-script,@0@'.format(meson.global_source_root() / 'build-aux' / 'test.map')
```
```suggestion
    '-Wl,-version-script,@0@'.format(meson.project_source_root() / 'build-aux' / 'test.map')
```

The two functions are equivalent except that global returns incorrect information if libdatrie is built as a subproject of another project -- it will then refer to the source root of the other project instead of libdatrie.

eli-schwartz commented on tools/meson.build:
```diff
@@ -11,5 +11,4 @@ trietool = executable(
 
-bindir_abs = get_option('prefix') + '/' + get_option('bindir')
-meson.add_install_script(sh, '-c', 'cd @0@ && ln -s trietool trietool-0.2 || true'.format(bindir_abs))
+meson.add_install_script(sh, '-c', 'cd ${MESON_INSTALL_DESTDIR_PREFIX}/@0@ && ln -s trietool trietool-0.2 || true'.format(get_option('bindir')))
```
```suggestion
install_symlink('trietool-0.2', pointing_to: 'trietool', install_dir: get_option('bindir))
```

whs commented:
Updated a bit to fix bad merge, and updated subproject usage to automatically override parent project dependency & trietool for better subproject UX.
